src/stubadub/core.cljc

```diff
@@ -1,29 +1,140 @@
+    (cond
+      has-return-value? return-value
+      has-return-fn? (return-value args)
+      :else nil)))
 
-(defmacro with-stub [func & opts-and-body]
+(deftype Stub [invocations opts]
+  #?@(:clj
+      [clojure.lang.IFn
+       (invoke [this]
+               (invoke-stub this []))
+       (invoke [this x1]
+               (invoke-stub this [x1]))
+       (invoke [this x1 x2]
+               (invoke-stub this [x1 x2]))
+       (invoke [this x1 x2 x3]
+               (invoke-stub this [x1 x2 x3]))
+       (invoke [this x1 x2 x3 x4]
+               (invoke-stub this [x1 x2 x3 x4]))
+       (invoke [this x1 x2 x3 x4 x5]
+               (invoke-stub this [x1 x2 x3 x4 x5]))
+       (invoke [this x1 x2 x3 x4 x5 x6]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20]
+               (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20]))
+       (invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 more]
```

Comment:
Shouldn't this be `& more` ?

Is performance in tests important enough to warrant all these arities, or could we do with a few and then go for `apply`?

Reply:
`more` is actually correct. Although `invoke` is an overloaded function, none of them are variadic. These functions are defined in [clojure.lang.IFn](https://clojure.github.io/clojure/javadoc/clojure/lang/IFn.html) and an implementation of each and every one of them is needed for the mock to work with any given arguments.

Reply:
Disregard my comment on `invoke` not being variadic. It definitely is, although this is the correct way to capture variadic methods in Java interfaces (i.e., they're already collected into an argument).

I moved the `IFn` boilerplate code to a macro, see `macros.clj`.